pkg/tnf/pkg/jobs/cronjob_controller.go

```diff
@@ -0,0 +1,249 @@
+		return fmt.Errorf("failed to ensure CronJob: %w", err)
+	}
+
+	return nil
+}
+
+func (c *CronJobController) ensureCronJob(ctx context.Context) error {
+	// Get existing CronJob if it exists
+	existing, err := c.kubeClient.BatchV1().CronJobs(c.namespace).Get(ctx, c.name, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get CronJob: %w", err)
+		}
+		// CronJob doesn't exist, create it
+		return c.createCronJob(ctx)
+	}
+
+	// CronJob exists, check if it needs updating
+	if c.needsUpdate(existing) {
+		klog.V(2).Infof("CronJob %s needs updating", c.name)
+		return c.updateCronJob(ctx, existing)
+	}
+
+	klog.V(4).Infof("CronJob %s is up to date", c.name)
+	return nil
+}
+
+func (c *CronJobController) createCronJob(ctx context.Context) error {
+	cronJob := c.buildCronJob()
+
+	_, err := c.kubeClient.BatchV1().CronJobs(c.namespace).Create(ctx, cronJob, metav1.CreateOptions{})
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			klog.V(4).Infof("CronJob %s already exists", c.name)
+			return nil
+		}
+		return fmt.Errorf("failed to create CronJob: %w", err)
+	}
+
+	klog.V(2).Infof("Created CronJob: %s/%s with schedule %s", c.namespace, c.name, c.schedule)
+	c.eventRecorder.Eventf("CronJobCreated", "Created CronJob: %s/%s", c.namespace, c.name)
+
+	return nil
+}
+
+func (c *CronJobController) updateCronJob(ctx context.Context, existing *batchv1.CronJob) error {
+	cronJob := c.buildCronJob()
+	cronJob.ResourceVersion = existing.ResourceVersion
+
+	_, err := c.kubeClient.BatchV1().CronJobs(c.namespace).Update(ctx, cronJob, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update CronJob: %w", err)
+	}
+
+	klog.V(2).Infof("Updated CronJob: %s/%s", c.namespace, c.name)
+	c.eventRecorder.Eventf("CronJobUpdated", "Updated CronJob: %s/%s", c.namespace, c.name)
+
+	return nil
+}
+
+func (c *CronJobController) buildCronJob() *batchv1.CronJob {
+	// Inject values into the template
+	injectedManifest := c.injectValues(c.manifest)
+
+	// Log the manifest before parsing to help debug YAML errors
+	klog.V(4).Infof("CronJob manifest before parsing:\n%s", string(injectedManifest))
+
+	// Parse the manifest with injected values
+	cronJob := ReadCronJobV1OrDie(injectedManifest)
+
+	return cronJob
+}
+
+// injectValues replaces <injected> placeholders in the manifest with actual values
+func (c *CronJobController) injectValues(manifest []byte) []byte {
+	result := string(manifest)
+
+	// Build command string for YAML array format
+	var commandBuilder strings.Builder
+	commandBuilder.WriteString("[")
+	for i, cmd := range c.command {
+		if i > 0 {
+			commandBuilder.WriteString(", ")
+		}
+		commandBuilder.WriteString(fmt.Sprintf(`"%s"`, cmd))
+	}
+	commandBuilder.WriteString("]")
+	commandStr := commandBuilder.String()
+
+	// Track which line we're on to do context-aware replacement
+	lines := strings.Split(result, "\n")
+	for i, line := range lines {
```

Comment:
tbh, I like using a `JobHookFunc` like in `runJobController()` much more than this string replacements... WDYT?

Reply:
I was confused as to how the JobHookFunc stuff worked :D
I was like - what is this sorcery

Reply:
Now that I am looking at it directly, it makes way more sense. Stealing this approach.